File: iosApp/iosApp/Features/Diary/notes/components/createnewnote/CreateNewNoteBottomSheet.swift
//
//  CreateNewNoteBottomSheet.swift
//  iosApp
//
//  Created by velkonost on 12.10.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK
import SwiftUIPager
import KMMViewModelSwiftUI
import KMPNativeCoroutinesAsync

struct AreaWrapper : Identifiable, Equatable, Hashable {
    var id: String = UUID().uuidString
    var area: Area
}

struct CreateNewNoteBottomSheet: View {
    
    @Binding private var state: CreateNewNoteViewState
    
    @State private var isAreaPickerVisible = false
    @State private var isSubNoteBlockVisible = false
    @State var currentAreaIndex: Int = 0
    
    let onAreaSelect: (Area) -> Void
    let onTextChanged: (String) -> Void
    let onPrivateChanged: () -> Void
    
    let onNewTagChanged: (String) -> Void
    let onAddNewTag: () -> Void
    let onTagDelete: (String) -> Void
    
    let onNewSubNoteChanged: (String) -> Void
    let onAddNewSubNote: () -> Void
    let onSubNoteDelete: (SubNoteUI) -> Void
    
    let onSetCompletionDate: (Int64?) -> Void
    
    let onCreateClick: () -> Void
    
    @State private var resourceMessageText: String?
    @State private var snackBar: MessageType.SnackBar?
    @State private var showSnackBar: Bool = false
    @State private var messageDequeObserver: Task<(), Error>? = nil
    
    init(
        state: Binding<CreateNewNoteViewState>,
        onAreaSelect: @escaping (Area) -> Void,
        onTextChanged: @escaping (String) -> Void,
        onPrivateChanged: @escaping () -> Void,
        onNewTagChanged: @escaping (String) -> Void,
        onAddNewTag: @escaping () -> Void,
        onTagDelete: @escaping (String) -> Void,
        onNewSubNoteChanged: @escaping (String) -> Void,
        onAddNewSubNote: @escaping () -> Void,
        onSubNoteDelete: @escaping (SubNoteUI) -> Void,
        onSetCompletionDate: @escaping (Int64?) -> Void,
        onCreateClick: @escaping () -> Void
    ) {
        self._state = state
        self.onAreaSelect = onAreaSelect
        self.onTextChanged = onTextChanged
        self.onPrivateChanged = onPrivateChanged
        
        self.onNewTagChanged = onNewTagChanged
        self.onAddNewTag = onAddNewTag
        self.onTagDelete = onTagDelete
        
        self.onNewSubNoteChanged = onNewSubNoteChanged
        self.onAddNewSubNote = onAddNewSubNote
        self.onSubNoteDelete = onSubNoteDelete
        
        self.onSetCompletionDate = onSetCompletionDate
        self.onCreateClick = onCreateClick
    }
    
    var body: some View {
        @State var isNotePrivate = state.isPrivate
        @State var newTag = state.newTag
        @State var newSubNote = state.newSubNote
        
        ZStack {
            Color.mainBackground
            
            if state.isLoading {
                VStack {
                    Loader()
                }
            } else {
                
                ScrollView(.vertical, showsIndicators: false) {
                    ScrollViewReader { value in
                        VStack {
                            Text(
                                state.type == NoteType.default_ ? SharedR.strings().create_note_title.desc().localized() : SharedR.strings().create_goal_title.desc().localized()
                            )
                            .style(.headlineSmall)
                            .foregroundColor(.textTitle)
                            .frame(alignment: .center)
                            
                            AreaPicker(
                                areas: state.availableAreas,
                                selectedArea: state.selectedArea,
                                noteType: state.type,
                                onAreaSelect: onAreaSelect,
                                isAreaPickerVisible: $isAreaPickerVisible
                            )
                            
                            MultilineTextField(
                                value: state.text,
                                placeholderText:
                                    state.type == NoteType.default_ ? SharedR.strings().create_note_text_hint.desc().localized() : SharedR.strings().create_goal_text_hint.desc().localized()
                            ) { value in
                                onTextChanged(value)
                            }
                            .onTapGesture {
                                withAnimation {
                                    value.scrollTo(1, anchor: .bottom)
                                }
                            }
                            .id(1)
                            
                            if state.type == NoteType.goal {
                                CompletionDateBlock(
                                    onSetCompletionDate: onSetCompletionDate
                                )
                            }
                            
                            PrivateSwitch(
                                onCheckedChange: onPrivateChanged,
                                isEnabled: state.selectedArea != nil && state.selectedArea?.isPrivate == false,
                                isPrivate: $isNotePrivate
                            )
                            
                            TagsBlock(
                                tags: state.tags,
                                newTag: $newTag,
                                onNewTagChanged: onNewTagChanged,
                                onAddNewTag: onAddNewTag,
                                onTagDelete: onTagDelete
                            )
                            .onTapGesture {
                                withAnimation {
                                    value.scrollTo(2, anchor: .bottom)
                                }
                            }
                            .id(2)
                            
                            if state.type == NoteType.goal {
                                SubNotesBlock(
                                    items: state.subNotes,
                                    newSubNote: $newSubNote,
                                    onNewSubNoteChanged: onNewSubNoteChanged,
                                    onAddNewSubNote: onAddNewSubNote,
                                    onSubNoteDelete: onSubNoteDelete,
                                    isSubNotesBlockPickerVisible: $isSubNoteBlockVisible
                                )
                                .onTapGesture {
                                    withAnimation {
                                        value.scrollTo(3, anchor: .bottom)
                                    }
                                }
                                .id(3)
                            }
                            
                            Spacer()
                            
                            
                        }
                        .padding(20)
                        .padding(.bottom, 140)
                    }
                    Spacer()
                }
            
            
            VStack {
                Spacer()
                
                VStack {
                    Spacer()
                        .frame(height: 20)
                    
                    AppButton(
                        labelText: SharedR.strings().diary_areas_create_button.desc().localized(),
                        isLoading: false
                    ) {
                        onCreateClick()
                    }
                    .padding(.bottom, 70)
                }
                .frame(minWidth: 0, maxWidth: .infinity)
                .background(
                    Rectangle()
                        .fill(LinearGradient(gradient: Gradient(colors: [
                            .mainBackground,
                            .mainBackground,
                            .mainBackground,
                            .clear
                        ]), startPoint: .bottom, endPoint: .top)
                        )
                    
                )
                
                
            }.ignoresSafeArea(.keyboard)
            
            Spacer()
        }
    }
        .ignoresSafeArea(.container)
        .onTapGesture {
            endTextEditing()
        }
        .snackBar(
            isShowing: $showSnackBar,
            text: resourceMessageText ?? "",
            snackBar: snackBar
        )
        .onAppear {
            if messageDequeObserver == nil {
                messageDequeObserver = Task {
                    for try await message in asyncSequence(for: MessageDeque.shared.invoke()) {
                        handle(resource: message)
                    }
                }
            }
        }
        .onDisappear {
            messageDequeObserver?.cancel()
            messageDequeObserver = nil
        }
}
}

extension CreateNewNoteBottomSheet {
    private func handle(resource message: Message) {
        switch message.messageType {
            
        case let snackBar as MessageType.SnackBar : do {
            if showSnackBar == false {
                resourceMessageText = message.text != nil ? message.text : message.textResource?.localized()
                self.snackBar = snackBar
                withAnimation {
                    showSnackBar.toggle()
                }
                DispatchQueue.main.asyncAfter(deadline: .now() + 5) {
                    Task { try await MessageDeque.shared.dequeue() }
                }
            }
        }
            
        default: break
        }
    }
}

File: androidApp/features/diary/src/main/kotlin/com/velkonost/getbetter/android/features/diary/notes/components/createnewnote/CreateNewNoteBottomSheet.kt
package com.velkonost.getbetter.android.features.diary.notes.components.createnewnote

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.ModalBottomSheetLayout
import androidx.compose.material.ModalBottomSheetState
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.velkonost.getbetter.core.compose.components.AppButton
import com.velkonost.getbetter.core.compose.components.Loader
import com.velkonost.getbetter.core.compose.components.MultilineTextField
import com.velkonost.getbetter.core.compose.components.note.areapicker.AreaPicker
import com.velkonost.getbetter.core.compose.components.note.completiondate.CompletionDateBlock
import com.velkonost.getbetter.core.compose.components.note.subnotes.SubNotesBlock
import com.velkonost.getbetter.core.compose.components.note.tags.TagsBlock
import com.velkonost.getbetter.core.compose.components.note.taskpicker.TaskPicker
import com.velkonost.getbetter.shared.core.model.area.Area
import com.velkonost.getbetter.shared.core.model.note.NoteType
import com.velkonost.getbetter.shared.core.model.task.TaskUI
import com.velkonost.getbetter.shared.core.model.ui.SubNoteUI
import com.velkonost.getbetter.shared.features.calendars.contracts.CreateNewNoteViewState
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.stringResource

@OptIn(ExperimentalMaterialApi::class, ExperimentalLayoutApi::class)
@Composable
fun CreateNewNoteBottomSheet(
    modifier: Modifier = Modifier,
    state: CreateNewNoteViewState,
    modalSheetState: ModalBottomSheetState,
    onAreaSelect: (Area) -> Unit,
    onTaskSelect: (TaskUI?) -> Unit,
    onTextChanged: (String) -> Unit,
    onPrivateChanged: (Boolean) -> Unit,
    onNewTagChanged: (String) -> Unit,
    onAddNewTag: () -> Unit,
    onTagDelete: (String) -> Unit,
    onNewSubNoteChanged: (String) -> Unit,
    onAddNewSubNote: () -> Unit,
    onSubNoteDelete: (SubNoteUI) -> Unit,
    onSetCompletionDate: (Long?) -> Unit,
    onCreateClick: () -> Unit
) {

    val isAreaPickerVisible = remember { mutableStateOf(false) }
    val isTaskPickerVisible = remember { mutableStateOf(false) }
    val isSubNotesBlockVisible = remember { mutableStateOf(false) }

    val scrollState = rememberScrollState()

    ModalBottomSheetLayout(
        sheetState = modalSheetState,
        sheetShape = RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp),
        sheetBackgroundColor = colorResource(resource = SharedR.colors.main_background),
        sheetContent = {
            if (state.isLoading) {
                Box(
                    modifier = modifier
                        .fillMaxWidth()
                        .fillMaxHeight(0.9f)
                        .padding(20.dp)
                ) {
                    Loader(modifier = modifier.align(Alignment.Center))
                }
            } else {
                Box(
                    modifier = modifier
                        .fillMaxWidth()
                        .fillMaxHeight(0.9f)
                ) {
                    Column(
                        modifier = modifier
                            .fillMaxWidth()
                            .fillMaxHeight(0.9f)
                            .padding(bottom = 40.dp)
                            .verticalScroll(scrollState)
                    ) {
                        Column(
                            modifier = modifier
                                .padding(20.dp)
                                .padding(bottom = 50.dp)
                        ) {
                            Text(
                                modifier = modifier.align(Alignment.CenterHorizontally),
                                text = stringResource(
                                    resource =
                                    if (state.type == NoteType.Default) SharedR.strings.create_note_title
                                    else SharedR.strings.create_goal_title
                                ),
                                color = colorResource(resource = SharedR.colors.text_title),
                                style = MaterialTheme.typography.headlineSmall
                            )

                            AreaPicker(
                                areas = state.availableAreas,
                                selectedArea = state.selectedArea,
                                isAreaPickerVisible = isAreaPickerVisible,
                                onAreaSelect = onAreaSelect,
                                modalSheetState = modalSheetState,
                                noteType = state.type
                            )

                            TaskPicker(
                                tasks = state.availableTasks,
                                selectedTask = state.selectedTask,
                                isTaskPickerVisible = isTaskPickerVisible,
                                modalSheetState = modalSheetState,
                                onTaskSelect = onTaskSelect
                            )

                            MultilineTextField(
                                value = state.text,
                                placeholderText = stringResource(
                                    resource =
                                    if (state.type == NoteType.Default) SharedR.strings.create_note_text_hint
                                    else SharedR.strings.create_goal_text_hint
                                ),
                                onValueChanged = { onTextChanged.invoke(it) }
                            )

                            AnimatedVisibility(visible = state.type == NoteType.Goal) {
                                CompletionDateBlock(
                                    onSetCompletionDate = onSetCompletionDate
                                )
                            }

                            PrivateSwitch(
                                isPrivate = state.isPrivate,
                                isEnable = state.selectedArea != null && state.selectedArea?.isPrivate == false,
                                onCheckedChange = onPrivateChanged
                            )

                            TagsBlock(
                                tags = state.tags,
                                newTag = state.newTag,
                                onNewTagChanged = onNewTagChanged,
                                onAddNewTag = onAddNewTag,
                                onTagDelete = onTagDelete
                            )

                            AnimatedVisibility(visible = state.type == NoteType.Goal) {
                                SubNotesBlock(
                                    items = state.subNotes,
                                    newSubNote = state.newSubNote,
                                    isSubNotesBlockVisible = isSubNotesBlockVisible,
                                    onAddNewSubNote = onAddNewSubNote,
                                    onNewSubNoteChanged = onNewSubNoteChanged,
                                    onSubNoteDelete = onSubNoteDelete
                                )
                            }

                        }
                    }

                    Column(
                        modifier = modifier.fillMaxWidth(),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Spacer(modifier = modifier.weight(1f))

                        Box(
                            modifier = modifier
                                .height(30.dp)
                                .fillMaxWidth()
                                .background(
                                    brush = Brush.verticalGradient(
                                        colors = listOf(
                                            Color.Transparent,
                                            colorResource(resource = SharedR.colors.main_background)
                                                .copy(alpha = 0.7f),
                                            colorResource(resource = SharedR.colors.main_background),
                                        ),
                                    ),
                                )
                        )

                        AppButton(
                            modifier = modifier
                                .align(Alignment.CenterHorizontally)
                                .padding(bottom = 70.dp),
                            labelText = stringResource(
                                resource = SharedR.strings.diary_areas_create_button
                            ),
                            isLoading = false,
                            onClick = onCreateClick
                        )
                    }

                }
            }
        }
    ) {

    }

    LaunchedEffect(modalSheetState.currentValue) {
        isAreaPickerVisible.value = false
        isSubNotesBlockVisible.value = false
    }
}